src/gui/src/histogramWidget.cpp

```diff
@@ -0,0 +1,242 @@
+///////////////////////////////////////////////////////////////////////////////
+// BSD 3-Clause License
+//
+// Copyright (c) 2019, The Regents of the University of California
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "histogramWidget.h"
+
+#include <QWidget>
+#include <QFrame>
+#include <QColor>
+#include <QString>
+#include <QtCharts>
+#include <vector>
+#include <cmath>
+#include <algorithm>
+#include <iostream>
+#include <set>
+
+#include <sta/MinMax.hh>
+#include "staGuiInterface.h"
+
+namespace gui {
+
+const int SELECT = 0;
+const int SLACK_MODE = 1;
+
+HistogramWidget::HistogramWidget(QWidget* parent)
+    : QDockWidget("Histogram Plotter", parent),
+      sta_(nullptr),
+      label_(new QLabel(this)),
+      mode_menu_(new QComboBox(this)),
+      chart_(new QChart),
+      display_(new QChartView(chart_)),
+      axis_x_(new QBarCategoryAxis(this)),
+      axis_y_(new QValueAxis(this))    
+{
+  setObjectName("histogram_widget"); // for settings
+
+  QWidget* container = new QWidget(this);
+  QVBoxLayout* layout = new QVBoxLayout;
+  QFrame* controls_frame = new QFrame;
+  QHBoxLayout* controls_layout = new QHBoxLayout;
+
+  mode_menu_->addItem("Select Mode");
+  mode_menu_->addItem("Endpoint Slack");
```

Comment:
As there is only one choice currently, can we default it to Endpoint Slack rather than forcing the user to select it.

Reply:
Theoretically, using the QComboBox signal `activated()` instead of `currentIndexChanged()`  should produce the desired effect, but I tried that way and what happens in reality is that the dropdown menu, despite having only one option, needs to be clicked either way... Is too bad to leave like this? If yes, I can remove the dropdown menu.

Reply:
Hopefully we'll have some more choices soon.  We can leave it for now.